setup.py

```diff
@@ -5,25 +5,43 @@
 from os import path
 from setuptools import setup
 
-_version_re = re.compile(r'__version__\s+=\s+(.*)')
+PACKAGE_NAME = "pythonwhat"
+REQUIREMENT_NAMES = ["markdown2", "jinja2", "asttokens", "dill", "numpy", "pandas"]
 
-PACKAGE_NAME = 'pythonwhat'
 HERE = path.abspath(path.dirname(__file__))
-with open(path.join(HERE, 'README.md'), encoding='utf-8') as fp:
+VERSION_FILE = path.join(HERE, PACKAGE_NAME, "__init__.py")
+REQUIREMENTS_FILE = path.join(HERE, "requirements.txt")
+README_FILE = path.join(HERE, "README.md")
+
+with open(VERSION_FILE, encoding="utf-8") as fp:
+    _version_re = re.compile(r"__version__\s+=\s+(.*)")
+    VERSION = str(ast.literal_eval(_version_re.search(fp.read()).group(1)))
+with open(REQUIREMENTS_FILE, encoding="utf-8") as fp:
+    req_txt = fp.read()
+    _requirements_re_template = r"^({}(?:\s*[<>=]+\s*\S*)?)\s*(?:#.*)?$"
+    REQUIREMENTS = [
+        re.search(_requirements_re_template.format(requirement), req_txt, re.M).group(0)
+        for requirement in REQUIREMENT_NAMES
+    ]
+with open(README_FILE, encoding="utf-8") as fp:
     README = fp.read()
-with open(path.join(HERE, PACKAGE_NAME, '__init__.py'), 'rb') as fp:
-    VERSION = str(ast.literal_eval(_version_re.search(
-        fp.read().decode('utf-8')).group(1)))
 
-setup(name=PACKAGE_NAME,
-      version=VERSION,
-      packages=[PACKAGE_NAME, 'pythonwhat.test_funcs'],
-      install_requires=["dill", "numpy", "pandas", "markdown2", "jinja2", "asttokens>=1.1.10"],
-      long_description=README,
-      long_description_content_type='text/markdown',
-      license='GNU version 3',
-      author='DataCamp',
-      author_email='content-engineering@datacamp.com',
-      maintainer='Filip Schouwenaars',
-      maintainer_email='filip@datacamp.com',
-      url='https://github.com/datacamp/pythonwhat')
+setup(
+    name=PACKAGE_NAME,
+    version=VERSION,
+    packages=[PACKAGE_NAME, "pythonwhat.test_funcs"],
+    install_requires=REQUIREMENTS,
```

Comment:
I would keep package dependencies simple!

Reading in requirements.txt seems like a dangerous practice. The  file contains a lot of packages that aren't necessary for running pythonwhat, and were tucked in to the unit tests (many when the package was first developed, before I started at DataCamp..).

install_requires should be the minimal set of packages required to run pythonwhat (otherwise, irrelevant packages will prevent its installation). I would keep it as a list, so it is simple and clear what the dependencies are, so they don't become artificially bloated.

see

https://packaging.python.org/discussions/install-requires-vs-requirements/#requirements-files

Reply:
It doesn't install them all, only the ones listed in this file. But there were version inconsistencies before, this way you control all versions in one place.